### DISCUSSION TITLE
acquireTriggersWithinLock default value if batchTriggerAcquisitionMaxCount > 1

### DISCUSSION THREAD

**Discussion by cn-byn** [quartznet/quartznet#1640] [Q&A] [answered]
Hi,

I was just looking into the source code and stumbled upon the following regarding locks.

Assuming we configure Quartz to be `clustered` and set both `threadCount` and `batchTriggerAcquisitionMaxCount` to something greater than 1. We don't set `acquireTriggersWithinLock`.

The documentation states about `quartz.jobStore.acquireTriggersWithinLock`: 

> If "quartz.scheduler.batchTriggerAcquisitionMaxCount" is set to > 1, and AdoJobStore is used, then this property must be set to "true" to avoid data corruption (as of Quartz 2 "true" is now the default if batchTriggerAcquisitionMaxCount is set > 1).`

However, looking at the code I think the last statement isn't completely correct because the lock is actually only acquired if there is more than one thread _currently_ available (`maxCount > 1`, see below).
If there is exactly one thread available at the moment, it doesn't acquire a lock, even though batchTriggerAcquisitionMaxCount is greater than 1.

```cs
// in QuartzSchedulerThread
var maxCount = Math.Min(availThreadCount, qsRsrcs.MaxBatchSize);
triggers = new List<IOperableTrigger>(await qsRsrcs.JobStore.AcquireNextTriggers(noLaterThan, maxCount, ...
```

```cs
// in JobStoreSupport.AcquireNextTriggers()
if (AcquireTriggersWithinLock || maxCount > 1)
{
    lockName = LockTriggerAccess;
}
else
{
    lockName = null;
}
```

I'm assuming this wouldn't be a problem without clustering, but couldn't this cause issues in clustered setups?

Thanks in advance.

**Comment by lahma** [comment, marked as answer]
The lock is only needed if more than one trigger will be acquired. So the > 1 will protect even if the locking hasn't been manually configured.
